### ISSUE TITLE
Low performance Crest 4.3 in HDRP 7.4.1

### ISSUE THREAD

**Issue by olavrv** [crest-ocean/crest#568]
**Describe the bug**
Crest LOD data takes 5ms + in GPU profiler. Was significantly lower on previous Crest versions

**Versions**

HDRP 7.4.1, Unity 2019.4.1f1, Crest v4.3

**Comment by daleeidd** [comment]
We have had a report of the depth cache running every frame (it should only run once on start) which will be fixed in 4.4. There is also a significant performance cost when underwater when using the _UnderwaterEnvironmentalLighting_ component. Is it either of these?

**Comment by olavrv** [comment]
It is neither. Also when looking straight up towards the sky, CrestLodData spikes and stays at around 10ms

**Comment by olavrv** [comment]
Just found that the it is the "QueryDisplacements" (GPU) that takes up 4.5ms of the total of 5ms

**Comment by huwb** [comment]
Thanks for creating this. Olav also reported on discord that he set the collision source to None (instead of compute shader queries) and the cost dropped to a fraction of a ms.

I gave a quick try on my laptop and am not seeing anything like this. Will contemplate..

**Comment by huwb** [comment]
@olavrv are you running with vsync enabled? If not, could you please enable it and see if it makes a difference?

The query system does not interact well with unlocked frame rate at the moment.

**Comment by olavrv** [comment]
@huwb vsync on/off has no effect. I also noticed that the line you suggested to change in the QueryBase.cs script was on line 460 in your script, but it was on line 439 in our script (we are using latest version on assetstore)

**Comment by olavrv** [comment]
To reproduce: download latest version on assetstore. Add gpu to profiler. Hit run, filter on "crestloddata" and see ms usage pr. frame. Unity 219.4.1f1 hdrp  7.4.1 . Use scene "boat" in crest examples.

**Comment by olavrv** [comment]
https://drive.google.com/file/d/1Zj-NTrtGyHUwxLovP2VvQ4gwmVzcJZpI/view?usp=sharing

**Comment by huwb** [comment]
I think theres a few things conflated here and we need to unpick a bit.

Firstly i tried on another GPU (GTX 1060), in editor, and saw a high cost attributed to CrestLodData. it jumps around a lot but is typically 5 or 6ms:

![image](https://user-images.githubusercontent.com/939901/86516502-0248b280-be19-11ea-8b17-df25fd9df08c.png)

Next i tried the same

**Comment by huwb** [comment]
@olavrv i just tried Crest HDRP 4.2 - in 2019.4.0f1 - and i see the same issue.

Crest HDRP 4.2 + 2019.3.9f1: same issue

I also have the feeling that this didnt occur in other versions, but i'm not able to prove that in my tests so far..

**Comment by olavrv** [comment]
@huwb I think it is important to get this GPU usage as low as possible. Crest has such a low performance cost otherwise...

**Comment by olavrv** [comment]
@huwb Do you know anything about the GPU usage of CrestLodData increasing when camera is pointing stright up towards the sky? I find this very strange (goes from 5ms to 10ms in the profiler)

**Comment by huwb** [comment]
No i dont - thats pretty odd. I need to try that out here, thanks for the reminder

**Comment by huwb** [comment]
I don't know what to do to progress this. Do you have ideas @daleeidd?

**Comment by daleeidd** [comment]
Unity freezes when I try to profile the GPU (even on built-in and latest Unity). So I am kind of stuck until I purchase new hardware :(

At least it appears to be only a serious issue in the editor.

**Comment by olavrv** [comment]
@huwb I can do a test in build, and see how much of an issue it is

**Comment by olavrv** [comment]
@huwb @daleeidd In a build, with just a crest ocean in the scene, CrestLodData uses around 4.5ms on the GPU (nvidia gtx 980). I find this is a bit much, and that it needs to be optimized. Also, is it possible to expose som sampling options (quality, what to sample etc), to make it possible for us to reduce the GPU cost according to individual needs? Large ships for example, don't need the same sample resolution as a small beach ball

**Comment by huwb** [comment]
Yeah I'm getting 0.8ms for crestloddata as I reported above, so something is wrong. I'll contemplate further..

**Comment by olavrv** [comment]
@huwb Hello! We are back from holliday, and tested a development build with the boat scene - on GTX 1080, it uses 5ms in editor, and 0.25 in build according to Nsight graphics. So I guess thats good?!

**Comment by huwb** [comment]
"great" :)

I'm still not sure how to make progress on this..

**Comment by olavrv** [comment]
@huwb We have been testing GPU performance on different stuff, and found one thing; CrestLODdata takes more performance in VR, and we wonder if this is related to the previous issue I mentioned about CrestLODdata taking less performance when camera looks up towards the sky. Should CrestLODdata GPU usage be related to what the camera sees?

**Comment by huwb** [comment]
No it should not. Everything under crestloddata should be entirely view independent..

**Comment by huwb** [comment]
Thinking about it, it might be that unity is failing to readback the query results asynchronously, and is stalling the GPU. That would match the symptoms i reckon. 

I'm not quite sure why unity would do that, or what we could do to avoid it. Perhaps theres some condition in which the async fails. Or maybe hdrp is already reading back a lot of data (?) and this is creating a bottleneck?
